syntax = "proto3";

package small.execution;

// =====================================================================
// small-db protobuf
// =====================================================================
import "src/type/type.proto";

service Insert {
  rpc Insert(Row) returns (UpsertResponse) {}
}

service Update {
  rpc Update(UpdateStmt) returns (UpsertResponse) {}
}

message Row {
  string table_name = 1;
  repeated string column_names = 2;
  repeated string column_values = 3;
}

message UpsertResponse {
  int64 affected_rows = 1;
}

// =====================================================================
// definition of sql AST, we don't use the proto buf along with pg_query
// since it will import pg_query.h and leads to error.
// (path: build/debug/_deps/libpg_query-src/protobuf/pg_query.proto)
// =====================================================================

message UpdateStmt {
  string table_name = 1;
  repeated SetClause set_clauses = 2;
  repeated Condition where_clauses = 3;
}

message SetClause {
  string column_name = 1;
  Expr target = 2;
}

message Expr {
  oneof expr {
    ColumnRef column_ref = 1;
    small.type.Datum literal = 2;
    MinusExpr minus_expr = 3;
  }
}

message ColumnRef {
  string column_name = 1;
}

message MinusExpr {
  Expr v1 = 1;
  Expr v2 = 2;
}

// Assume the condition is always equality for simplicity.
message Condition {
  string column_name = 1;
  Expr value = 2;
}
